File: system-test/dotnet/Dsl.Core/UseCase/UseCases/ViewOrderVerification.cs
using Dsl.Core.Shared;
using Driver.Port.Dtos;
using Shouldly;
using System.Globalization;
using Common;

namespace Dsl.Core.UseCase.UseCases;

public class ViewOrderVerification : ResponseVerification<ViewOrderResponse>
{
    public ViewOrderVerification(ViewOrderResponse response, UseCaseContext context)
        : base(response, context)
    {
    }

    public ViewOrderVerification OrderNumber(string orderNumberResultAlias)
    {
        var expectedOrderNumber = Context.GetResultValue(orderNumberResultAlias);
        Response.OrderNumber.ShouldBe(expectedOrderNumber,
            $"Expected order number to be '{expectedOrderNumber}', but was '{Response.OrderNumber}'");
        return this;
    }

    public ViewOrderVerification Sku(string skuParamAlias)
    {
        var expectedSku = Context.GetParamValue(skuParamAlias);
        Response.Sku.ShouldBe(expectedSku,
            $"Expected SKU to be '{expectedSku}', but was '{Response.Sku}'");
        return this;
    }

    public ViewOrderVerification Quantity(int expectedQuantity)
    {
        Response.Quantity.ShouldBe(expectedQuantity,
            $"Expected quantity: {expectedQuantity}, but got: {Response.Quantity}");
        return this;
    }

    public ViewOrderVerification Quantity(string expectedQuantity)
    {
        return Quantity(Converter.ToInteger(expectedQuantity)!.Value);
    }

    public ViewOrderVerification Status(OrderStatus expectedStatus)
    {
        Response.Status.ShouldBe(expectedStatus,
            $"Expected status: {expectedStatus}, but got: {Response.Status}");
        return this;
    }

    public ViewOrderVerification Status(string expectedStatus)
    {
        return Status(Enum.Parse<OrderStatus>(expectedStatus));
    }

    public ViewOrderVerification Country(string expectedCountry)
    {
        Response.Country.ShouldBe(expectedCountry,
            $"Expected country: '{expectedCountry}', but got: '{Response.Country}'");
        return this;
    }

    public ViewOrderVerification UnitPrice(decimal expectedUnitPrice)
    {
        Response.UnitPrice.ShouldBe(expectedUnitPrice,
            $"Expected unit price: {expectedUnitPrice}, but got: {Response.UnitPrice}");
        return this;
    }

    public ViewOrderVerification UnitPrice(string expectedUnitPrice)
    {
        return UnitPrice(Converter.ToDecimal(expectedUnitPrice)!.Value);
    }

    public ViewOrderVerification UnitPriceGreaterThanZero()
    {
        Response.UnitPrice.ShouldBeGreaterThan(0m,
            $"Unit price should be positive, but was: {Response.UnitPrice}");
        return this;
    }

    public ViewOrderVerification SubtotalPrice(decimal expectedSubtotalPrice)
    {
        Response.SubtotalPrice.ShouldBe(expectedSubtotalPrice,
            $"Expected subtotal price: {expectedSubtotalPrice}, but got: {Response.SubtotalPrice}");
        return this;
    }

    public ViewOrderVerification SubtotalPrice(string expectedSubtotalPrice)
    {
        return SubtotalPrice(Converter.ToDecimal(expectedSubtotalPrice)!.Value);
    }

    public ViewOrderVerification SubtotalPriceGreaterThanZero()
    {
        Response.SubtotalPrice.ShouldBeGreaterThan(0m,
            $"Subtotal price should be positive, but was: {Response.SubtotalPrice}");
        return this;
    }

    public ViewOrderVerification DiscountRateGreaterThanOrEqualToZero()
    {
        Response.DiscountRate.ShouldBeGreaterThanOrEqualTo(0m,
            $"Discount rate should be non-negative, but was: {Response.DiscountRate}");
        return this;
    }

    public ViewOrderVerification DiscountAmountGreaterThanOrEqualToZero()
    {
        Response.DiscountAmount.ShouldBeGreaterThanOrEqualTo(0m,
            $"Discount amount should be non-negative, but was: {Response.DiscountAmount}");
        return this;
    }

    public ViewOrderVerification TaxRateGreaterThanOrEqualToZero()
    {
        Response.TaxRate.ShouldBeGreaterThanOrEqualTo(0m,
            $"Tax rate should be non-negative, but was: {Response.TaxRate}");
        return this;
    }

    public ViewOrderVerification TaxAmountGreaterThanOrEqualToZero()
    {
        Response.TaxAmount.ShouldBeGreaterThanOrEqualTo(0m,
            $"Tax amount should be non-negative, but was: {Response.TaxAmount}");
        return this;
    }

    public ViewOrderVerification TotalPriceGreaterThanZero()
    {
        Response.TotalPrice.ShouldBeGreaterThan(0m,
            $"Total price should be positive, but was: {Response.TotalPrice}");
        return this;
    }

    public ViewOrderVerification BasePrice(decimal expectedBasePrice)
    {
        Response.BasePrice.ShouldBe(expectedBasePrice,
            $"Expected base price to be {expectedBasePrice}, but was {Response.BasePrice}");
        return this;
    }

    public ViewOrderVerification BasePrice(string expectedBasePrice)
    {
        return BasePrice(Converter.ToDecimal(expectedBasePrice)!.Value);
    }

    public ViewOrderVerification BasePriceGreaterThanZero()
    {
        Response.BasePrice.ShouldBeGreaterThan(0m,
            $"Base price should be positive, but was: {Response.BasePrice}");
        return this;
    }

    public ViewOrderVerification DiscountRate(decimal expectedDiscountRate)
    {
        Response.DiscountRate.ShouldBe(expectedDiscountRate,
            $"Expected discount rate to be {expectedDiscountRate}, but was {Response.DiscountRate}");
        return this;
    }

    public ViewOrderVerification DiscountRate(string expectedDiscountRate)
    {
        return DiscountRate(Converter.ToDecimal(expectedDiscountRate)!.Value);
    }

    public ViewOrderVerification DiscountAmount(decimal expectedDiscountAmount)
    {
        Response.DiscountAmount.ShouldBe(expectedDiscountAmount,
            $"Expected discount amount to be {expectedDiscountAmount}, but was {Response.DiscountAmount}");
        return this;
    }

    public ViewOrderVerification DiscountAmount(string expectedDiscountAmount)
    {
        return DiscountAmount(Converter.ToDecimal(expectedDiscountAmount)!.Value);
    }

    public ViewOrderVerification TaxRate(decimal expectedTaxRate)
    {
        Response.TaxRate.ShouldBe(expectedTaxRate,
            $"Expected tax rate to be {expectedTaxRate}, but was {Response.TaxRate}");
        return this;
    }

    public ViewOrderVerification TaxRate(string expectedTaxRate)
    {
        return TaxRate(Converter.ToDecimal(expectedTaxRate)!.Value);
    }

    public ViewOrderVerification TaxAmount(decimal expectedTaxAmount)
    {
        Response.TaxAmount.ShouldBe(expectedTaxAmount,
            $"Expected tax amount to be {expectedTaxAmount}, but was {Response.TaxAmount}");
        return this;
    }

    public ViewOrderVerification TaxAmount(string expectedTaxAmount)
    {
        return TaxAmount(Converter.ToDecimal(expectedTaxAmount)!.Value);
    }

    public ViewOrderVerification TotalPrice(decimal expectedTotalPrice)
    {
        Response.TotalPrice.ShouldBe(expectedTotalPrice,
            $"Expected total price to be {expectedTotalPrice}, but was {Response.TotalPrice}");
        return this;
    }

    public ViewOrderVerification TotalPrice(string expectedTotalPrice)
    {
        return TotalPrice(Converter.ToDecimal(expectedTotalPrice)!.Value);
    }

    public ViewOrderVerification AppliedCouponCode(string expectedCouponCodeAlias)
    {
        var expectedCouponCode = Context.GetParamValue(expectedCouponCodeAlias);
        Response.AppliedCouponCode.ShouldBe(expectedCouponCode,
            $"Expected applied coupon code to be '{expectedCouponCode}', but was '{Response.AppliedCouponCode}'");
        return this;
    }

    public ViewOrderVerification OrderNumberHasPrefix(string expectedPrefix)
    {
        Response.OrderNumber.ShouldStartWith(expectedPrefix, Case.Sensitive,
            $"Expected order number to start with '{expectedPrefix}', but was '{Response.OrderNumber}'");
        return this;
    }
}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/usecases/ViewOrderVerification.java
package com.mycompany.myshop.testkit.dsl.core.usecase.usecases;

import com.mycompany.myshop.testkit.driver.port.dtos.ViewOrderResponse;
import com.mycompany.myshop.testkit.domainvaluetypes.OrderStatus;
import com.mycompany.myshop.testkit.dsl.core.shared.ResponseVerification;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.common.Converter;

import java.math.BigDecimal;
import java.time.Instant;

import static org.assertj.core.api.Assertions.assertThat;

public class ViewOrderVerification extends ResponseVerification<ViewOrderResponse> {
    public ViewOrderVerification(ViewOrderResponse response, UseCaseContext context) {
        super(response, context);
    }

    public ViewOrderVerification orderNumber(String orderNumberResultAlias) {
        var expectedOrderNumber = getContext().getResultValue(orderNumberResultAlias);
        var actualOrderNumber = getResponse().getOrderNumber();
        assertThat(actualOrderNumber)
                .withFailMessage("Expected order number to be '%s', but was '%s'", expectedOrderNumber, actualOrderNumber)
                .isEqualTo(expectedOrderNumber);
        return this;
    }

    public ViewOrderVerification sku(String skuParamAlias) {
        var expectedSku = getContext().getParamValue(skuParamAlias);
        var actualSku = getResponse().getSku();
        assertThat(actualSku)
                .withFailMessage("Expected SKU to be '%s', but was '%s'", expectedSku, actualSku)
                .isEqualTo(expectedSku);
        return this;
    }

    public ViewOrderVerification quantity(int expectedQuantity) {
        var actualQuantity = getResponse().getQuantity();
        assertThat(actualQuantity)
                .withFailMessage("Expected quantity to be %d, but was %d", expectedQuantity, actualQuantity)
                .isEqualTo(expectedQuantity);
        return this;
    }

    public ViewOrderVerification quantity(String expectedQuantity) {
        return quantity(Converter.toInteger(expectedQuantity));
    }

    public ViewOrderVerification unitPrice(double expectedUnitPrice) {
        var expectedPrice = Converter.toBigDecimal(expectedUnitPrice);
        var actualPrice = getResponse().getUnitPrice();
        assertThat(actualPrice)
                .withFailMessage("Expected unit price to be %s, but was %s", expectedPrice, actualPrice)
                .isEqualByComparingTo(expectedPrice);
        return this;
    }

    public ViewOrderVerification unitPrice(String expectedUnitPrice) {
        return unitPrice(Converter.toDouble(expectedUnitPrice));
    }

    public ViewOrderVerification status(OrderStatus expectedStatus) {
        var actualStatus = getResponse().getStatus();
        assertThat(actualStatus)
                .withFailMessage("Expected status to be %s, but was %s", expectedStatus, actualStatus)
                .isEqualTo(expectedStatus);
        return this;
    }

    public ViewOrderVerification status(String expectedStatus) {
        return status(OrderStatus.valueOf(expectedStatus));
    }

    public ViewOrderVerification totalPrice(BigDecimal expectedTotalPrice) {
        var totalPrice = getResponse().getTotalPrice();
        assertThat(totalPrice)
                .withFailMessage("Expected total price to be %s, but was %s", expectedTotalPrice, totalPrice)
                .isEqualByComparingTo(expectedTotalPrice);
        return this;
    }

    public ViewOrderVerification totalPrice(double expectedTotalPrice) {
        return totalPrice(Converter.toBigDecimal(expectedTotalPrice));
    }

    public ViewOrderVerification totalPrice(String expectedTotalPrice) {
        return totalPrice(Converter.toBigDecimal(expectedTotalPrice));
    }

    public ViewOrderVerification totalPriceGreaterThanZero() {
        var totalPrice = getResponse().getTotalPrice();
        assertThat(totalPrice)
                .withFailMessage("Total price should be positive, but was: %s", totalPrice)
                .isGreaterThan(BigDecimal.ZERO);
        return this;
    }

    public ViewOrderVerification subtotalPrice(BigDecimal expectedSubtotalPrice) {
        var actualSubtotalPrice = getResponse().getSubtotalPrice();
        assertThat(actualSubtotalPrice)
                .withFailMessage("Expected subtotal price to be %s, but was %s", expectedSubtotalPrice, actualSubtotalPrice)
                .isEqualByComparingTo(expectedSubtotalPrice);
        return this;
    }

    public ViewOrderVerification subtotalPrice(double expectedSubtotalPrice) {
        return subtotalPrice(Converter.toBigDecimal(expectedSubtotalPrice));
    }

    public ViewOrderVerification subtotalPrice(String expectedSubtotalPrice) {
        return subtotalPrice(Converter.toDouble(expectedSubtotalPrice));
    }

    public ViewOrderVerification basePrice(BigDecimal expectedBasePrice) {
        var actualBasePrice = getResponse().getBasePrice();
        assertThat(actualBasePrice)
                .withFailMessage("Expected base price to be %s, but was %s", expectedBasePrice, actualBasePrice)
                .isEqualByComparingTo(expectedBasePrice);
        return this;
    }

    public ViewOrderVerification basePrice(double expectedBasePrice) {
        return basePrice(Converter.toBigDecimal(expectedBasePrice));
    }

    public ViewOrderVerification basePrice(String expectedBasePrice) {
        return basePrice(Converter.toBigDecimal(expectedBasePrice));
    }

    public ViewOrderVerification discountAmount(BigDecimal expectedDiscountAmount) {
        var actualDiscountAmount = getResponse().getDiscountAmount();
        assertThat(actualDiscountAmount)
                .withFailMessage("Expected discount amount to be %s, but was %s", expectedDiscountAmount, actualDiscountAmount)
                .isEqualByComparingTo(expectedDiscountAmount);
        return this;
    }

    public ViewOrderVerification discountAmount(double expectedDiscountAmount) {
        return discountAmount(Converter.toBigDecimal(expectedDiscountAmount));
    }

    public ViewOrderVerification discountAmount(String expectedDiscountAmount) {
        return discountAmount(Converter.toBigDecimal(expectedDiscountAmount));
    }

    public ViewOrderVerification taxRate(BigDecimal expectedTaxRate) {
        var actualTaxRate = getResponse().getTaxRate();
        assertThat(actualTaxRate)
                .withFailMessage("Expected tax rate to be %s, but was %s", expectedTaxRate, actualTaxRate)
                .isEqualByComparingTo(expectedTaxRate);
        return this;
    }

    public ViewOrderVerification taxRate(double expectedTaxRate) {
        return taxRate(Converter.toBigDecimal(expectedTaxRate));
    }

    public ViewOrderVerification taxRate(String expectedTaxRate) {
        return taxRate(Converter.toBigDecimal(expectedTaxRate));
    }

    public ViewOrderVerification taxAmount(BigDecimal expectedTaxAmount) {
        var actualTaxAmount = getResponse().getTaxAmount();
        assertThat(actualTaxAmount)
                .withFailMessage("Expected tax amount to be %s, but was %s", expectedTaxAmount, actualTaxAmount)
                .isEqualByComparingTo(expectedTaxAmount);
        return this;
    }

    public ViewOrderVerification taxAmount(double expectedTaxAmount) {
        return taxAmount(Converter.toBigDecimal(expectedTaxAmount));
    }

    public ViewOrderVerification taxAmount(String expectedTaxAmount) {
        return taxAmount(Converter.toBigDecimal(expectedTaxAmount));
    }

    public ViewOrderVerification discountRate(BigDecimal expectedDiscountRate) {
        var actualDiscountRate = getResponse().getDiscountRate();
        assertThat(actualDiscountRate)
                .withFailMessage("Expected discount rate to be %s, but was %s", expectedDiscountRate, actualDiscountRate)
                .isEqualByComparingTo(expectedDiscountRate);
        return this;
    }

    public ViewOrderVerification discountRate(double expectedDiscountRate) {
        return discountRate(Converter.toBigDecimal(expectedDiscountRate));
    }

    public ViewOrderVerification appliedCouponCode(String expectedCouponCodeAlias) {
        var expectedCouponCode = getContext().getParamValue(expectedCouponCodeAlias);
        var actualCouponCode = getResponse().getAppliedCouponCode();
        assertThat(actualCouponCode)
                .withFailMessage("Expected applied coupon code to be '%s', but was '%s'", expectedCouponCode, actualCouponCode)
                .isEqualTo(expectedCouponCode);
        return this;
    }

    public void orderNumberHasPrefix(String expectedPrefix) {
        var actualOrderNumber = getResponse().getOrderNumber();
        assertThat(actualOrderNumber)
                .withFailMessage("Expected order number to start with '%s', but was: %s", expectedPrefix, actualOrderNumber)
                .startsWith(expectedPrefix);
    }

    public ViewOrderVerification orderTimestamp(Instant expectedTimestamp) {
        var actualTimestamp = getResponse().getOrderTimestamp();
        assertThat(actualTimestamp)
                .withFailMessage("Expected order timestamp to be %s, but was %s", expectedTimestamp, actualTimestamp)
                .isEqualTo(expectedTimestamp);
        return this;
    }

    public ViewOrderVerification orderTimestamp(String expectedTimestamp) {
        return orderTimestamp(Converter.toInstant(expectedTimestamp));
    }

    public ViewOrderVerification orderTimestampIsNotNull() {
        var actualTimestamp = getResponse().getOrderTimestamp();
        assertThat(actualTimestamp)
                .withFailMessage("Expected order timestamp to be set, but was null")
                .isNotNull();
        return this;
    }

}
